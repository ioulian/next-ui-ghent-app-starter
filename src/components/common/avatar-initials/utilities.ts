import { escapeSVG } from "@/utils/svg";

/**
 * Will create a simple image with some text in it. Useful when you don't have any image for an user
 *
 * @param initials String to show in the image
 * @param backgroundColor
 * @param textColor
 * @returns svg string
 */
export const getAvatarInitials = (
  initials: string,
  backgroundColor: string = "currentColor",
  textColor: string = "#fff",
): string => {
  return escapeSVG(
    `<svg width="512" height="512" xmlns="http://www.w3.org/2000/svg"><rect height="512" width="512" y="-1" x="-1" fill="${backgroundColor}"/><text transform="matrix(1 0 0 1 0 0)" font-weight="bold" stroke="#000" xml:space="preserve" font-family="'Inter', sans-serif" font-size="180" stroke-width="0" x="50%" y="54%" dominant-baseline="middle" text-anchor="middle" fill="${textColor}">${initials.toLocaleUpperCase()}</text></svg>`,
  );
};
