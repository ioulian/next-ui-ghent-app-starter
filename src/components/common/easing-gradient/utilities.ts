import { easingCoordinates } from "easing-coordinates";
import { mix, parse, ColorSpace, sRGB, Lab, serialize } from "colorjs.io/fn";

// If you want to use more color spaces, you'll need to register them first
ColorSpace.register(sRGB);
ColorSpace.register(Lab);

/**
 * Will generate a smooth gradient to be used in css
 *
 * Some info: https://non-boring-gradients.netlify.app/
 * "Ported" from https://github.com/larsenwork/postcss-easing-gradients
 *
 * @param from Start color
 * @param to End color
 * @param easing Easing function
 * @param stops Number of stops to use. Do not go overboard with it!
 * @returns Comma separated color stops to be used in "linear-gradient"
 */
export const createEasingGradient = (
  from: string,
  to: string,
  easing: string = "cubic-bezier(0.5, 0, 0.5, 1)",
  stops: number = 16,
) => {
  const coordinates = easingCoordinates(easing, stops);

  // TS ignore is there because the types for serialize are not correct. This works.
  return coordinates
    .map(
      ({ x, y }) =>
        `${serialize(
          mix(parse(from), parse(to), y, {
            space: "lab",
            outputSpace: "srgb",
          }),
        )} ${Math.abs(x * 100).toFixed(2)}%`,
    )
    .join(", ");
};
