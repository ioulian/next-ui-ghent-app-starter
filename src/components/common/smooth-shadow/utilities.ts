// "Ported" from: https://smoothshadows.com/
import { easingCoordinates } from "easing-coordinates";
import Color from "colorjs.io";
import { parse, ColorSpace, sRGB } from "colorjs.io/fn";

ColorSpace.register(sRGB);

export interface SmoothShadowOptionsType {
  startAlpha: number;
  endAlpha: number;
  startBlur: number;
  endBlur: number;
  layers: number;
  x: number;
  y: number;
  color: string;
  alphaEasing: string;
  blurEasing: string;
  distanceEasing: string;
}

/**
 * Create a smooth shadow composed of different "normal" shadows in css.
 *
 * @returns String to be used in css `box-shadow`
 */
export const createSmoothShadow = ({
  /**
   * Start opacity of the shadow
   */
  startAlpha,

  /**
   * End opacity of the shadow
   */
  endAlpha,

  /**
   * Start blur  of the shadow
   */
  startBlur,

  /**
   * End blur  of the shadow
   */
  endBlur,

  /**
   * Number of "normal" css shadows to generate. More is more smooth, but will tank performance quite fast
   */
  layers,
  x,
  y,
  color,
  alphaEasing,
  blurEasing,
  distanceEasing,
}: SmoothShadowOptionsType): string => {
  const distanceCoordinates = easingCoordinates(distanceEasing, layers);
  const blurCoordinates = easingCoordinates(blurEasing, layers);
  const alphaCoordinates = easingCoordinates(alphaEasing, layers);
  const parsedColor = new Color(parse(color));
  const rgb: [number, number, number] = [
    Math.floor(parsedColor.srgb.r * 255),
    Math.floor(parsedColor.srgb.g * 255),
    Math.floor(parsedColor.srgb.b * 255),
  ];

  return Array(layers)
    .fill(0)
    .map((_, i) => ({
      x: distanceCoordinates[i].y * x,
      y: distanceCoordinates[i].y * y,
      blur: startBlur + blurCoordinates[i].y * (endBlur - startBlur),
      alpha: startAlpha + alphaCoordinates[i].y * (endAlpha - startAlpha),
    }))
    .map(
      ({ x, y, blur, alpha }) =>
        `${x.toFixed(1)}px ${y.toFixed(1)}px ${blur.toFixed(
          2,
        )}px rgba(${rgb.join(", ")}, ${alpha.toFixed(2)})`,
    )
    .join(", ");
};
