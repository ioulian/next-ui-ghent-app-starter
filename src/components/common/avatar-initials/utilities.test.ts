/* eslint-disable i18next/no-literal-string */

import { getAvatarInitials } from "./utilities";

describe("utilities", () => {
  it("generates a correct svg", () => {
    const value = getAvatarInitials("TE");
    expect(value).toMatchSnapshot();
  });
});
