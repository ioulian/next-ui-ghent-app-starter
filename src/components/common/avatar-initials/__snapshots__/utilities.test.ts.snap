// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utilities generates a correct svg 1`] = `"data:image/svg+xml;charset=UTF-8,<svg width="512" height="512" xmlns="http://www.w3.org/2000/svg"><rect height="512" width="512" y="-1" x="-1" fill="currentColor"/><text transform="matrix(1 0 0 1 0 0)" font-weight="bold" stroke="%23000" xml:space="preserve" font-family="'Inter', sans-serif" font-size="180" stroke-width="0" x="50%" y="54%" dominant-baseline="middle" text-anchor="middle" fill="%23fff">TE</text></svg>"`;
